src/XMakeBuildEngine/Definition/Project.cs

```diff
@@ -1289,98 +1286,59 @@ private int ItemMatchesInSpecCompareViaExpander(string itemToMatch, string itemS
 
             if (expandedMatches > nonExpandedMatches)
             {
-                // return the number of occurences when properties AND items are expanded to get the correct occurence count
-
-                // return the provenance WITHOUT item expansion. Otherwise the items coming from a referenced item get interpreted as StringLiteral
-                // include="*.cs;@(Compile)" needs to return Inconclusive|Glob and not Inconclusive|Glob|StringLiteral
-                Provenance provenanceFromExpandedProperties;
-                ItemMatchesInSpec(itemToMatch, expand(itemSpec, ExpanderOptions.ExpandProperties), out provenanceFromExpandedProperties);
+                // return the number of occurences when properties are expanded to get the correct occurence count
 
                 provenance = Provenance.Inconclusive | provenanceFromExpandedProperties;
                 return expandedMatches;
             }
             else
             {
-                provenance = provenanceFromNonExpandedString;
+                provenance = provenanceFromNonExpandedProperties;
                 return nonExpandedMatches;
             }
         }
 
-        private int ItemMatchesInSpec(string itemToMatch, string itemSpec, out Provenance provenance)
+        private int ItemMatchesInSpec(string itemToMatch, ItemSpec<ProjectProperty, ProjectItem> itemSpec, out Provenance provenance)
         {
             provenance = Provenance.Undefined;
 
             var occurrences = 0;
 
-            if (string.IsNullOrEmpty(itemSpec))
-            {
-                return occurrences;
-            }
-
-            foreach (var itemFragment in ExpressionShredder.SplitSemiColonSeparatedList(itemSpec))
+            var fragmentsMatchingItem = itemSpec.FragmentsMatchingItem(itemToMatch, out occurrences);
+            foreach (var fragment in fragmentsMatchingItem)
             {
-                if (IsPropertyReferenceFragment(itemFragment) || IsItemReferenceFragment(itemFragment))
+                if (fragment is ValueFragment)
                 {
-                    continue;
+                    provenance |= Provenance.StringLiteral;
                 }
-
-                if (IsGlobFragment(itemFragment) && ItemMatchesGlob(itemFragment, itemToMatch))
+                else if (fragment is GlobFragment)
                 {
                     provenance |= Provenance.Glob;
-                    occurrences++;
                 }
-
-                if (ItemMatchesStringLiteral(itemToMatch, itemFragment))
+                else if (fragment is ItemExpressionFragment<ProjectProperty, ProjectItem>)
                 {
-                    provenance |= Provenance.StringLiteral;
-                    occurrences++;
+                    provenance |= Provenance.Inconclusive;
+                }
+                else
+                {
+                    throw new NotImplementedException();
```

Comment:
This is scary. Are there current scenarios where this could get hit, or is it just for possible future expansion?

(side note: this will all be so much clearer with C# 7.0!)